src/components/Trainer/VirtualCube.tsx

```diff
@@ -0,0 +1,144 @@
+import { CheckIcon, CloseIcon, EditIcon, CopyIcon, TriangleUpIcon, TriangleDownIcon } from "@chakra-ui/icons"
+import { Box, Button, FormControl, FormErrorMessage, Heading, IconButton, Input, Stack, Spinner, Text, Center, useColorModeValue, Spacer, Flex, Container, Switch } from "@chakra-ui/react"
+import { useAtom } from "jotai"
+import { atomWithStorage } from "jotai/utils"
+import { Suspense, useEffect, useRef, useState } from "react"
+import { isFaceMove, moveSeqToString, parseNotation, parseVC, replaceBadApostrophes, simplifyMoves } from "src/lib"
+import type { Mask, Move, MoveSeq } from "src/lib/types"
+import Cube from "../Cube/Cube"
+import TrainerCard from "./common/TrainerCard"
+
+interface VirtualCubeProps {
+  scramble: MoveSeq
+  solution: MoveSeq
+  setSolution: (newSolution: MoveSeq) => void
+  mask?: Mask
+  showEO?: boolean
+}
+
+const VCInputAtom = atomWithStorage<boolean>("vc-input", false)
+
+// TODO: show "Failed to generate solution" message if solution === null
+export default function VirtualCube({ scramble, solution, setSolution, mask, showEO }: VirtualCubeProps) {
+  const [isEditing, setEditing] = useState(false)
+  const [inputSolution, setInputSolution] = useState("")
+  const [showCube, setShowCube] = useState(false)
+  // const inputIsInvalid = !isValidNotation(inputSolution)
+  // TODO: make solver handle solutions that change cube orientation
+  // for now, only HTM solutions are ok
+  const inputIsInvalid = inputSolution !== "" && !inputSolution.trim().split(" ").every(token => isFaceMove(token as Move))
+
+  const cubeBackground = useColorModeValue("gray.200", undefined)
+
+  const [VCInput, setVCInput] = useAtom(VCInputAtom)
+
+  // reset the input whenever a new solution is set
+  useEffect(() => {
+    setInputSolution("")
+  }, [solution])
+
+  const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = e => {
+    if(VCInput){
+      const inputLength = e.target.value.length
+      const normalNotation = e.target.value.substring(0, inputLength-1)
+      const newChar = e.target.value.charAt(inputLength-1)
+      const newMove = parseVC(newChar) + ((parseVC(newChar))?" ":"")
+      setInputSolution(normalNotation + newMove)
+    } else {
+      const filteredInputSolution = replaceBadApostrophes(e.target.value)
+      setInputSolution(filteredInputSolution)
+    }
+  }
+
+  const submitSolution = () => {
+    if (!inputIsInvalid) {
+      setSolution(simplifyMoves(parseNotation(inputSolution)))
+      setEditing(false)
+    }
+  }
+
+  const handleFormSubmit: React.FormEventHandler<HTMLFormElement> = e => {
+    e.preventDefault()
+    submitSolution()
+  }
+
+  const copySolution = () => {
+    navigator.clipboard.writeText(solution.join(" "))
+  }
+
+  return (
+    <TrainerCard>
+      <Flex minWidth='max-content' direction="row" alignItems='center' gap='2'>
+        <Heading size="md">virtual cube</Heading>
```

Comment:
I'm wondering if there's a better name for this. Maybe "solution editor"? Both this section and the "solutions" section have a virtual cube

Reply:
I was going for SolutionEditor in the beginning, so yeah